### PR TITLE
refactor(pubsub): split SubscriptionSession [2]

### PR DESCRIPTION
**Pull request by coryan**
Copy the functionality to split message batches and split them into
individual messages. For now, this class is only used in tests. A future
PR will replace `pubsub_internal::SubscriptionSession` with a class that
uses this new component.

Part of the work for #4970

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4992)

### CODE REVIEW

File: google/cloud/pubsub/internal/subscription_message_source.h [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+ */
+class SubscriptionMessageSource {
+ public:
+  virtual ~SubscriptionMessageSource() = default;
+
+  /// Start the source, set up the callback. Calling multiple times should have
+  /// no effect, only the first callback is used.
+  virtual void Start(MessageCallback) = 0;
+
+  /// Shutdown the source, cancel any outstanding requests and or timers, no
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/pubsub/internal/subscription_message_source.h]: Fixed.

File: google/cloud/pubsub/internal/subscription_message_source.h [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+  /// Request more messages from the source.
+  virtual void Read(std::size_t max_callbacks) = 0;
+
+  /**
+   * Positive acknowledgement the message associated with @p ack_id.
+   *
+   * The application has successfully handled this message and no new deliveries
+   * are necessary. The @p size parameter should be the original message size
+   * estimate, it may be used by the message source to flow control large
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/pubsub/internal/subscription_message_source.h]: Named the parameter explicitly, here and it `SubscriptionBatchSource`.

File: google/cloud/pubsub/internal/subscription_message_source.h [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+   * messages.
+   */
+  virtual future<Status> AckMessage(std::string const& ack_id,
+                                    std::size_t size) = 0;
+
+  /**
+   * Negative acknowledgement for message associated with @p ack_id.
+   *
+   * The application has not able to handle this message, the service should
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/pubsub/internal/subscription_message_source.h]: Fixed (I think).

File: google/cloud/pubsub/internal/subscription_message_source.h [hunk 4]
```diff
@@ -56,20 +56,20 @@ class SubscriptionMessageSource {
...
 
   /**
    * Negative acknowledgement for message associated with @p ack_id.
    *
-   * The application has not able to handle this message, the service should
-   * re-deliver the message (subject to the topic and subscription
-   * configuration). The @p size parameter should be the original message size
-   * estimate, it may be used by the message source to flow control large
-   * messages.
+   * The application has not able to handle this message. Nacking a message
```
Comment (devbww) [inline google/cloud/pubsub/internal/subscription_message_source.h:68]: s/has not able/has not been able/?

Comment (coryan) [reply google/cloud/pubsub/internal/subscription_message_source.h]: Doh!

**Pull request merged**